### PR TITLE
Tic-Tac_Toe Game Logic

### PR DESCRIPTION
**Pull request by Paul-Clue**
## What we did:
- We modularized the main file.
- We cleaned up all unnecessary comments.
- We created a new class in the interface.rb file.

### CODE REVIEW

File: bin/main.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  mark = game.p_one_loop(mark, o_wins, player_one.name)
+  p 'run check!'
+  game.updated_board(mark, player_one.mark)
+
+  puts "\n"
+  puts "\n"
+  moves += 1
+  game.update_array
+
+  # Check The Score
```
Comment (rindrajosia) [inline bin/main.rb]: No inline comment

File: bin/main.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  end
+
+  game.updated_board(mark, player_two.mark)
+
+  puts "\n"
+  puts "\n"
+  moves += 1
+  game.update_array
+
+  # Check The Score
```
Comment (rindrajosia) [inline bin/main.rb]: No inline comment

File: lib/interface.rb [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+      next unless score == 1
+
+      game_on = false
+      x_wins = true
+      break
+    end
+
+    puts '------------------------'
+    puts '------------------------'
+    # Columns
```
Comment (rindrajosia) [inline lib/interface.rb]: No inline comment

File: lib/interface.rb [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      next unless score == 1
+
+      game_on = false
+      o_wins = true
+      break
+    end
+
+    puts '------------------------'
+    puts '------------------------'
+    # Columns
```
Comment (rindrajosia) [inline lib/interface.rb]: No inline comment

File: lib/interface.rb [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    mark = mak
+    o_wins = o_w
+    player_one = p_one
+    until $char_array.include? mark
+      break unless o_wins == false
+
+      puts 'You can only select an integer, and that integer must be on the game board.'
+      puts "\n"
+      puts "#{player_one} your turn. Enter the number of where you would like to place your mark. "
+      mark = gets.chomp
```
Comment (rindrajosia) [inline lib/interface.rb]: No `puts/print` and `gets.chomp` outside main.rb file

File: lib/interface.rb [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      score += 1 if v == %w[x x x]
+      next unless score == 1
+
+      game_on = false
+      x_wins = true
+      break
+    end
+
+    puts '------------------------'
+    puts '------------------------'
```
Comment (rindrajosia) [inline lib/interface.rb]: No `puts/print` and `gets.chomp` outside main.rb file

File: lib/interface.rb [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+      score += 1 if v == %w[o o o]
+      next unless score == 1
+
+      game_on = false
+      o_wins = true
+      break
+    end
+
+    puts '------------------------'
+    puts '------------------------'
```
Comment (rindrajosia) [inline lib/interface.rb]: No `puts/print` and `gets.chomp` outside main.rb file

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**CHANGES REQUIRED**
HELLO!! You did a great job in this millstone very nice work but there are some changes that you have to make, please check below:

REQUIRED:
- [ ] No `puts/print` and `gets.chomp` outside `main.rb` file.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).
- [ ] No inline comments left after the previous milestone. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/no_comment.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).

REQUIRED: (Behaviour)
- [ ] In your game, if `player two` choose an `invalid input or an occupied position`, the application just informs him that it is an invalid input and `asks directly player one` **to choose a position**. Please fix it.




- [ ]  Use attr_readers/attr_writers. Only use attr_accessors if you really need it.

After addressing those minor bugs submit another CRR.

Happy Coding!
Rindra Josia

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required ♻🛑**

<hr>

Hello there, It's a great job you've done so far in this project and it looks professional. However, there are still some small changes to make before moving on!

- [ ] Remove .gitkeep file from the lib directory.
- [ ] When a player makes an invalid move, the game should inform them of that move, and ask them to try again i.e. when I enter number 2 on the board and it's already taken, the game should tell me so.

<hr>

## **[OPTIONAL]**

- Think about dividing your interface.rb file into two files each contain one class i.e. player.rb and board.rb (this way your code would be more professional and organized)

<hr>

Good luck and happy coding 👨‍💻💻

**Comment by Paul-Clue** [comment]
We just erased the .gitkeep file. However, we think that the error message that is in the game is sufficient for instructing the player that he or she has entered some invalid input. The message says: "You can only select an integer, and that integer must be on the game board." This message advises that only integers are valid entries and that the integer must be present on the game board.

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello team! 👋 
Good job so far! 👍 
But there are a few things to do to get the project approved 💪 
Kindly check my comments below 👇  and resubmit your project for a new code review once done ⏩ 
Good luck! 😺 
## Mandatory changes
- I still can see .gitkeep in the repo, kindly remove it
- When the player 2 introduces an invalid move the message 'You can only select an integer, and that integer must be on the game board.' appears during all the game for all the player 2 moves, kindly make it appear only when appropriate
- Kindly add different messages, one for an invalid input, one for a cell already taken

**Review by LamiaSristy — CHANGES_REQUESTED** [review]
# Status: Change Required. 🔴 

Hello Team, Your game has some issues to fix. 🙌

Please check my points here: 👇

`[REQUIRED]`

- [ ]  For the diagonal position [3,5,7] your game does not count as a win, please check the image below:
win
- [ ]  If the last/9th move is not a win then it's a draw, please check the image below:
scene1

`[OPTIONAL]`

- [ ]  you have used attr_accessors, consider using only attr_readers/attr_writers.
- [ ]  The messages also not accurate in your game, please rewrite them.


Please make the changes and ask for a re-review.

Feel free to reach me at @lamiasristy 🙋

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Status: Change Required.
- Hello Team, I can see that you are working really hard on this project. :muscle: 
 ### [Important]
- Kindly fix the issues raised by @LamiaSristy before you submit another review.
### Do not make any changes in the rubocop file - they represent style guidelines that you share with your team - which is a group of all Microverse students.

- Feel free to contact me on slack [Oloo Moses], in case of any doubts or concerns.

**Comment by ixboy** [comment]
we have made the changes required by the last code reviewer.  would you please test our code again ?

we have completed all the requirements...  Thank You

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED!! :recycle: 
Hello team!! :wave:  excellent job working on the changes 
I can see that there is a confusion with the last review, there's only one problem that you don't fix. see it bellow
------

- [ ] When you are playing as 'o' and you made a winning move, the game ask the next player 'x' to made a move, when the player made the move it shows that the winner is the 'x' player. 

Screenshot from 2021-01-20 21-05-28

As you can see the game should finish when the 'B' player choose the 7. 

Screenshot from 2021-01-20 21-12-19

In this last image, you can see that is telling you the wrong winner.

- [ ] Don't change the original rubocop.yml file. 
-----------------------------
Work on these changes and ask for another review :muscle: 
You are working really hard on this project great job, just made this little changes to get an excellent project. :clap: 
HAPPY CODING!!

**Comment by Paul-Clue** [comment]
This is a link to a video where I show that our code is working properly: https://www.loom.com/share/95cffc59bd0f442a9afe8d0edb882318

Please watch this video because you guys keep writing back to us that our code is not working, but as you will see in the video it is working fine.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**Changes required**
Hello team!!
You need just to fix one issue to get your project approved, I am sure that it is not going to take much time.

- [ ] When _player one chooses 'o'_, and when he makes a winning move, the game still _continues to ask player 2 to enter a number_ instead of telling **that player one is the winner.** 

Please check this video: [link](https://www.loom.com/share/a98aefbed8604decbff738f42085ade5).

Kindly fix It and submit a new code review request
Thanks
Rindra Josia

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required 🛑**

<hr>

Hello Team, you've done a great job working on this project. However, there is a very small issue you need to do to make the game work properly.

- [ ] The game is breaking at the 9th move even if it's a win
Inkedmain rb - tictactoe - Visual Studio Code 1_21_2021 6_30_20 AM_LI

- [ ] "Small Hint" please look at the following image and fix line 74 this is the one causing the above issue ```break if moves >= 9```
main rb - tictactoe - Visual Studio Code 1_21_2021 6_41_34 AM (2)_LI

**Review by MahmoudBakr23 — APPROVED** [review]
APPROVED 🎉💯👍

<hr>

Thanks a lot for the effort you've made in this project and in the requested changes.
You can merge and submit. 
Good luck with the next ❤️

**Pull request merged**